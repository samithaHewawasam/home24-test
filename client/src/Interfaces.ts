export interface Price {
  currency: string
  regular: {
    value: number
  }
}
export interface Image {
  path: string
}
export interface Article {
  name: string
  index: number
  prices: Price
  images: Image[]
}

export interface ChildCategory {
  name: string
  urlPath: string
}

export interface Category {
  name: string
  articleCount: number
  childrenCategories: { list: ChildCategory[] }
  categoryArticles: { articles: Article[] }
}

export interface State<T> {
  data: T
  isError: boolean
  isLoading: boolean
}
