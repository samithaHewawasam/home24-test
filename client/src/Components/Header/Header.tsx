import { ShoppingCartOutlined } from '@ant-design/icons'
import { AutoComplete, Badge, Input, PageHeader, Radio } from 'antd'
import { querySuggestions } from 'GraphRepository'
import { useGQLRequest } from 'Http'
import { Article, State } from 'Interfaces'
import PropTypes from 'prop-types'
import React, { memo } from 'react'
import { useNavigate } from 'react-router-dom'

interface Suggestion {
  name: string
  image: string
}

const intialData: State<{ suggestions: Suggestion[] }> = {
  data: {
    suggestions: [
      {
        name: '',
        image: ''
      }
    ]
  },
  isError: false,
  isLoading: false
}

export const Header: React.FC<{
  setLocale: (e: string) => void
  cartItems: Article[]
}> = memo(({ setLocale, cartItems }) => {
  const navigate = useNavigate()
  const { data, isError, isLoading, setQuery } = useGQLRequest<{
    suggestions: Suggestion[]
  }>(intialData)

  const handleSearch = (searchText: string): void => {
    if (searchText === null) return

    if (typeof setQuery !== 'undefined') {
      setQuery({
        query: querySuggestions,
        variables: { locale: 'de_DE', format: 'WEBP', prefix: searchText }
      })
    }
  }

  const handleSelect = (selectedText: string): void => {
    navigate(`/search?query=${selectedText}`)
  }

  return (
    <PageHeader
      ghost={false}
      title="Home24"
      subTitle="The online destination for home and living."
      extra={[
        <AutoComplete
          key={'searchAutoComplete'}
          style={{ width: 300 }}
          options={data?.suggestions.map(({ name }) => ({ value: name }))}
          dropdownMatchSelectWidth
          notFoundContent={
            isError === true ? 'Something Wrong here' : 'Not Found'
          }
          autoFocus
          onSearch={(text: string) => handleSearch(text)}
          onSelect={(text: string) => handleSelect(text)}
        >{console.log(window.location.pathname)}
          <Input.Search
            size="middle"
            placeholder="Search here"
            enterButton
            loading={isLoading}
          />
        </AutoComplete>,
        <Radio.Group
          defaultValue="de_DE"
          buttonStyle="solid"
          key={'locale'}
          onChange={(e) => setLocale(e.target.value)}
        >
          <Radio.Button value="de_DE">DE</Radio.Button>
          <Radio.Button value="fr_FR">FR</Radio.Button>
        </Radio.Group>,
        <Badge
          className="site-badge-count-109"
          count={cartItems.length}
          key={'shopingCart'}
          style={{ backgroundColor: '#52c41a' }}
        >
          <ShoppingCartOutlined style={{ fontSize: '32px', color: '#08c' }} />
        </Badge>
      ]}
    ></PageHeader>
  )
})

Header.propTypes = {
  cartItems: PropTypes.array.isRequired,
  setLocale: PropTypes.func.isRequired
}

Header.displayName = 'Header'
export default Header
