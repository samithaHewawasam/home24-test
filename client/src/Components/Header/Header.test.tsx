import { cleanup, render } from '@testing-library/react'
import { act } from '@testing-library/react-hooks/dom'
import { Article } from 'Interfaces'
import React from 'react'
import { BrowserRouter } from 'react-router-dom'
import renderer from 'react-test-renderer'
import Header from './Header'

const cartItems: Article[] = [
  {
    name: 'Beistelltisch Hoya (2er-Set)',
    index: 1,
    prices: {
      currency: 'EUR',
      regular: {
        value: 11499
      }
    },
    images: [
      {
        path: 'https://cdn1.home24.net/images/media/catalog/product/200x200/png/-/1/-1000174115-190524-15511900114-IMAGE-P000000001000174115.webp'
      }
    ]
  }
]

describe('Header', () => {
  const HeaderComponent = (
    <BrowserRouter>
      <Header cartItems={cartItems} setLocale={jest.fn} />
    </BrowserRouter>
  )
  beforeEach(async () => {
    await act(async () => {
      render(HeaderComponent)
    })
  })

  afterEach(() => {
    cleanup()
  })

  /**
   * @description SnapShot for Header
   */
  it('renders Header correctly', () => {
    const tree = renderer.create(HeaderComponent).toJSON()
    expect(tree).toMatchSnapshot()
  })

  it('render the Header Title', () => {
    const _HeaderTitle = document.querySelector(
      '.ant-page-header-heading-title'
    )?.innerHTML
    expect(_HeaderTitle).toContain('Home24')
  })

  it('render the Header Sub Title', () => {
    const _HeaderSubTitle = document.querySelector(
      '.ant-page-header-heading-sub-title'
    )?.innerHTML
    expect(_HeaderSubTitle).toContain(
      'The online destination for home and living.'
    )
  })
})
