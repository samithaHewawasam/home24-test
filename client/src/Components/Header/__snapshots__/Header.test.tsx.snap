// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Header renders Header correctly 1`] = `
<div
  className="ant-page-header"
>
  <div
    className="ant-page-header-heading"
  >
    <div
      className="ant-page-header-heading-left"
    >
      <span
        className="ant-page-header-heading-title"
        title="Home24"
      >
        Home24
      </span>
      <span
        className="ant-page-header-heading-sub-title"
        title="The online destination for home and living."
      >
        The online destination for home and living.
      </span>
    </div>
    <span
      className="ant-page-header-heading-extra"
    >
      <div
        className="ant-space ant-space-horizontal ant-space-align-center"
        style={Object {}}
      >
        <div
          className="ant-space-item"
          style={
            Object {
              "marginRight": 8,
            }
          }
        >
          <div
            className="ant-select ant-select-auto-complete ant-select-single ant-select-customize-input ant-select-show-search"
            onBlur={[Function]}
            onFocus={[Function]}
            onKeyDown={[Function]}
            onKeyUp={[Function]}
            onMouseDown={[Function]}
            style={
              Object {
                "width": 300,
              }
            }
          >
            <div
              className="ant-select-selector"
              onClick={[Function]}
              onMouseDown={[Function]}
            >
              <span
                className="ant-select-selection-search"
              >
                <span
                  className="ant-input-group-wrapper ant-input-search ant-input-search-middle ant-input-search-with-button ant-select-selection-search-input"
                  style={
                    Object {
                      "opacity": null,
                    }
                  }
                >
                  <span
                    className="ant-input-wrapper ant-input-group"
                  >
                    <input
                      aria-activedescendant="rc_select_TEST_OR_SSR_list_0"
                      aria-autocomplete="list"
                      aria-controls="rc_select_TEST_OR_SSR_list"
                      aria-haspopup="listbox"
                      aria-owns="rc_select_TEST_OR_SSR_list"
                      autoComplete="off"
                      autoFocus={true}
                      className="ant-input"
                      hidden={null}
                      id="rc_select_TEST_OR_SSR"
                      onBlur={[Function]}
                      onChange={[Function]}
                      onCompositionEnd={[Function]}
                      onCompositionStart={[Function]}
                      onFocus={[Function]}
                      onKeyDown={[Function]}
                      onMouseDown={[Function]}
                      onPaste={[Function]}
                      placeholder="Search here"
                      readOnly={false}
                      role="combobox"
                      style={null}
                      type="search"
                      unselectable={null}
                      value=""
                    />
                    <span
                      className="ant-input-group-addon"
                    >
                      <button
                        className="ant-btn ant-btn-primary ant-input-search-button"
                        disabled={false}
                        onClick={[Function]}
                        onMouseDown={[Function]}
                        type="button"
                      >
                        <span
                          aria-label="search"
                          className="anticon anticon-search"
                          role="img"
                        >
                          <svg
                            aria-hidden="true"
                            data-icon="search"
                            fill="currentColor"
                            focusable="false"
                            height="1em"
                            viewBox="64 64 896 896"
                            width="1em"
                          >
                            <path
                              d="M909.6 854.5L649.9 594.8C690.2 542.7 712 479 712 412c0-80.2-31.3-155.4-87.9-212.1-56.6-56.7-132-87.9-212.1-87.9s-155.5 31.3-212.1 87.9C143.2 256.5 112 331.8 112 412c0 80.1 31.3 155.5 87.9 212.1C256.5 680.8 331.8 712 412 712c67 0 130.6-21.8 182.7-62l259.7 259.6a8.2 8.2 0 0011.6 0l43.6-43.5a8.2 8.2 0 000-11.6zM570.4 570.4C528 612.7 471.8 636 412 636s-116-23.3-158.4-65.6C211.3 528 188 471.8 188 412s23.3-116.1 65.6-158.4C296 211.3 352.2 188 412 188s116.1 23.2 158.4 65.6S636 352.2 636 412s-23.3 116.1-65.6 158.4z"
                            />
                          </svg>
                        </span>
                      </button>
                    </span>
                  </span>
                </span>
              </span>
              <span
                className="ant-select-selection-placeholder"
              />
            </div>
          </div>
        </div>
        <div
          className="ant-space-item"
          style={
            Object {
              "marginRight": 8,
            }
          }
        >
          <div
            className="ant-radio-group ant-radio-group-solid"
          >
            <label
              className="ant-radio-button-wrapper ant-radio-button-wrapper-checked"
            >
              <span
                className="ant-radio-button ant-radio-button-checked"
                style={Object {}}
              >
                <input
                  checked={true}
                  className="ant-radio-button-input"
                  onBlur={[Function]}
                  onChange={[Function]}
                  onFocus={[Function]}
                  onKeyDown={[Function]}
                  onKeyPress={[Function]}
                  onKeyUp={[Function]}
                  type="radio"
                  value="de_DE"
                />
                <span
                  className="ant-radio-button-inner"
                />
              </span>
              <span>
                DE
              </span>
            </label>
            <label
              className="ant-radio-button-wrapper"
            >
              <span
                className="ant-radio-button"
                style={Object {}}
              >
                <input
                  checked={false}
                  className="ant-radio-button-input"
                  onBlur={[Function]}
                  onChange={[Function]}
                  onFocus={[Function]}
                  onKeyDown={[Function]}
                  onKeyPress={[Function]}
                  onKeyUp={[Function]}
                  type="radio"
                  value="fr_FR"
                />
                <span
                  className="ant-radio-button-inner"
                />
              </span>
              <span>
                FR
              </span>
            </label>
          </div>
        </div>
        <div
          className="ant-space-item"
          style={Object {}}
        >
          <span
            className="ant-badge site-badge-count-109"
          >
            <span
              aria-label="shopping-cart"
              className="anticon anticon-shopping-cart"
              role="img"
              style={
                Object {
                  "color": "#08c",
                  "fontSize": "32px",
                }
              }
            >
              <svg
                aria-hidden="true"
                data-icon="shopping-cart"
                fill="currentColor"
                focusable="false"
                height="1em"
                viewBox="0 0 1024 1024"
                width="1em"
              >
                <path
                  d="M922.9 701.9H327.4l29.9-60.9 496.8-.9c16.8 0 31.2-12 34.2-28.6l68.8-385.1c1.8-10.1-.9-20.5-7.5-28.4a34.99 34.99 0 00-26.6-12.5l-632-2.1-5.4-25.4c-3.4-16.2-18-28-34.6-28H96.5a35.3 35.3 0 100 70.6h125.9L246 312.8l58.1 281.3-74.8 122.1a34.96 34.96 0 00-3 36.8c6 11.9 18.1 19.4 31.5 19.4h62.8a102.43 102.43 0 00-20.6 61.7c0 56.6 46 102.6 102.6 102.6s102.6-46 102.6-102.6c0-22.3-7.4-44-20.6-61.7h161.1a102.43 102.43 0 00-20.6 61.7c0 56.6 46 102.6 102.6 102.6s102.6-46 102.6-102.6c0-22.3-7.4-44-20.6-61.7H923c19.4 0 35.3-15.8 35.3-35.3a35.42 35.42 0 00-35.4-35.2zM305.7 253l575.8 1.9-56.4 315.8-452.3.8L305.7 253zm96.9 612.7c-17.4 0-31.6-14.2-31.6-31.6 0-17.4 14.2-31.6 31.6-31.6s31.6 14.2 31.6 31.6a31.6 31.6 0 01-31.6 31.6zm325.1 0c-17.4 0-31.6-14.2-31.6-31.6 0-17.4 14.2-31.6 31.6-31.6s31.6 14.2 31.6 31.6a31.6 31.6 0 01-31.6 31.6z"
                />
              </svg>
            </span>
            <sup
              className="ant-scroll-number ant-badge-count"
              data-show={true}
              style={
                Object {
                  "backgroundColor": "#52c41a",
                }
              }
              title={1}
            >
              <span
                className="ant-scroll-number-only"
                onTransitionEnd={[Function]}
                style={
                  Object {
                    "transition": "none",
                  }
                }
              >
                <span
                  className="ant-scroll-number-only-unit current"
                >
                  1
                </span>
              </span>
            </sup>
          </span>
        </div>
      </div>
    </span>
  </div>
</div>
`;
